### PR TITLE
gadget: add validation for "$kernel:ref" style content

### PR DESCRIPTION
**Pull request by mvo5**
This commit adds validation for the new "$kernel:ref" style
references. It's relatively restricted right now we can open
up as needed.

Split out from https://github.com/snapcore/snapd/pull/9149

### CODE REVIEW

File: gadget/validate.go [hunk 1]
```diff
@@ -216,6 +219,15 @@ func validateVolumeContentsPresence(gadgetSnapRootDir string, vol *LaidOutVolume
 			continue
 		}
 		for _, c := range s.Content {
+			if strings.HasPrefix(c.UnresolvedSource, "$kernel:") {
+				// This only validates that the ref is valid.
+				// Resolving happens with ResolveContentPaths()
+				match := assetsKernelRefRE.FindStringSubmatch(c.UnresolvedSource)
```
Comment (bboozzoo) [inline gadget/validate.go:225]: Had a quick look at some validation of socket path and layouts and I'm wondering if we need a regex, or maybe something as simple as:

```
p := c.UnresolvedSource[len("$kernel:"):]
likelyOk :=  filepath.Clean(p) == p &&  !strings.Contains(p, "..") &&  !filepath.IsAbs(p)
```
and then let the filesystem handle the rest.

Comment (bboozzoo) [reply gadget/validate.go:225]: Nvm, looked at #9777 totally forgot the format is slightly different. Maybe regex is actually ok.

File: gadget/validate_test.go [hunk 2]
```diff
@@ -392,3 +394,42 @@ func (s *validateGadgetTestSuite) TestValidateEncryptionSupportHappy(c *C) {
...
+		{"$kernel:aa/bb", true},
+		{"$kernel:a-a/b-b", true},
+		{"$kernel:aB-0/cD-3", true},
+		{"$kernel:aB-0/foo-21B.dtb", true},
+		{"$kernel:aB-0/nested/bar-77A.raw", true},
+		// no starting with "-"
+		{"$kernel:-/-", false},
+		// assets and content need to be there
+		{"$kernel:/", false},
+		{"$kernel:a/", false},
```
Comment (anonymouse64) [inline gadget/validate_test.go:423]:
```suggestion
		{"$kernel:a/", false},
		{"$kernel:/a", false},
```

**Review by anonymouse64 — APPROVED** [review]
LGTM, one additional test case that would be good I think

File: gadget/validate.go [hunk 3]
```diff
@@ -208,6 +209,8 @@ func ensureSystemSaveRuleConsistency(state *validationState) error {
 
 // content validation
 
+var assetsKernelRefRE = regexp.MustCompile(`^\$kernel:([a-zA-Z0-9]+[a-zA-Z0-9-]*)/([a-zA-Z0-9/]+[a-zA-Z0-9/.-]*)$`)
```
Comment (bboozzoo) [inline gadget/validate.go:212]: The following will be considered valid:
```
$kernel:foo///
$kernel:foo/a../
$kernel:foo/a-/
```

**Review by mvo5 — COMMENTED** [review]
Thanks @bboozzoo for your improvements! :+1:

File: gadget/validate.go [hunk 4]
```diff
@@ -208,6 +209,39 @@ func ensureSystemSaveRuleConsistency(state *validationState) error {
 
 // content validation
 
+// XXX: needs to match kernel/kernel.go
+var validKernelAssetName = regexp.MustCompile("^[a-zA-Z0-9][a-zA-Z0-9-]*$")
```
Comment (pedronis) [inline gadget/validate.go:213]: shouldn't we import it from there then?  or a Validate* functions?

File: gadget/validate.go [hunk 5]
```diff
@@ -216,6 +250,14 @@ func validateVolumeContentsPresence(gadgetSnapRootDir string, vol *LaidOutVolume
...
 		for _, c := range s.Content {
+			if strings.HasPrefix(c.UnresolvedSource, "$kernel:") {
+				// This only validates that the ref is valid.
+				// Resolving happens with ResolveContentPaths()
+				if _, _, err := splitKernelRef(c.UnresolvedSource); err != nil {
+					return fmt.Errorf("cannot use kernel reference %q: %v", c.UnresolvedSource, err)
+				}
+				continue
+			}
 			// TODO: detect and skip Content with "$kernel:" style refs if there is no kernelSnapRootDir passed in as well
```
Comment (pedronis) [inline gadget/validate.go:261]: feels like this TODO will still need addressing later, and so it needs to be moved before the new if

Comment (bboozzoo) [reply gadget/validate.go]: done

File: gadget/validate.go (hunk 4 above)
Comment (bboozzoo) [reply gadget/validate.go]: done

**Review by anonymouse64 — APPROVED** [review]
the new changes still lgtm

File: gadget/validate.go [hunk 6]
```diff
@@ -208,6 +209,36 @@ func ensureSystemSaveRuleConsistency(state *validationState) error {
...
+	// sane path
+	if !strings.HasPrefix(kernelRef, "$kernel:") {
+		return "", "", fmt.Errorf("not a kernel ref")
+	}
+	assetAndContent := kernelRef[len("$kernel:"):]
+	idx := strings.Index(assetAndContent, "/")
+	if idx == -1 {
+		return "", "", fmt.Errorf("invalid asset and content in kernel ref %q", kernelRef)
+	}
+	asset, content = assetAndContent[:idx], assetAndContent[idx+1:]
```
Comment (pedronis) [inline gadget/validate.go:224]: can't this code be a strings.SplitN ?

File: gadget/validate.go [hunk 7]
```diff
@@ -208,6 +209,36 @@ func ensureSystemSaveRuleConsistency(state *validationState) error {
 
 // content validation
 
+func splitKernelRef(kernelRef string) (asset, content string, err error) {
+	// kernel ref has format: $kernel:<asset-name>/<content-name> where
```
Comment (pedronis) [inline gadget/validate.go:213]: should content-name be more like content-path ?

File: gadget/validate_test.go [hunk 8]
```diff
@@ -392,3 +394,59 @@ func (s *validateGadgetTestSuite) TestValidateEncryptionSupportHappy(c *C) {
...
+		{source: "$kernel:#garbage/a"},
+		// invalid content part
+		{source: "$kernel:a//"},
+		{source: "$kernel:a///"},
+		{source: "$kernel:a////"},
+		{source: "$kernel:a/a/../"},
+	} {
+		gadgetYaml := strings.Replace(gadgetYamlContentKernelRef, "REPLACE_WITH_TC", tc.source, -1)
+		makeSizedFile(c, filepath.Join(s.dir, "meta/gadget.yaml"), 0, []byte(gadgetYaml))
+		err := gadget.Validate(s.dir, nil, nil)
```
Comment (pedronis) [inline gadget/validate_test.go:438]: this needs a master merge, this is now ValidateContent I think and the signature has changed a bit

**Review by pedronis — CHANGES_REQUESTED** [review]
this needs a master merge because of the recent changes. some other comments, I haven't quite finished the review because atm the test can't be run

File: gadget/validate.go [hunk 9]
```diff
@@ -257,6 +258,36 @@ func ensureSystemSaveRuleConsistency(state *validationState) error {
...
+	// asset name and content is listed in kernel.yaml, content looks like a
+	// sane path
+	if !strings.HasPrefix(kernelRef, "$kernel:") {
+		return "", "", fmt.Errorf("not a kernel ref")
```
Comment (pedronis) [inline gadget/validate.go:266]: this case doesn't seem tested

File: gadget/validate.go (hunk 6 above)
Comment (mvo5) [reply gadget/validate.go:273]: Sure, updated the code.

**Review by pedronis — APPROVED** [review]
thanks

**Pull request merged**